ethexe/tx-pool/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2024 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Ethexe transaction pool.
+
+mod validation;
+
+#[cfg(test)]
+mod tests;
+
+pub use ethexe_common::tx_pool::{RawTransacton, SignedTransaction, Transaction};
+
+use anyhow::{Context as _, Result};
+use ethexe_db::Database;
+use ethexe_signer::{Address, Signature, ToDigest};
+use futures::{
+    ready,
+    stream::{FusedStream, Stream},
+};
+use gprimitives::H160;
+use parity_scale_codec::Encode;
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::sync::mpsc;
+use validation::TxValidator;
+
+/// Transaction pool service.
+///
+/// Serves as an interface for the transaction pool core.
+pub struct TxPoolService {
+    db: Database,
+    // No concurrent access for ready_tx is here,
+    // so no need for the mutex.
+    ready_tx: VecDeque<SignedTransaction>,
+    readiness_sender: mpsc::UnboundedSender<()>,
+    readiness_receiver: mpsc::UnboundedReceiver<()>,
+}
+
+impl TxPoolService {
+    pub fn new(db: Database) -> Self {
+        let (readiness_sender, readiness_receiver) = mpsc::unbounded_channel();
+        Self {
+            db,
+            ready_tx: VecDeque::new(),
+            readiness_sender,
+            readiness_receiver,
+        }
+    }
+
+    /// Basically validates the transaction and includes the transaction
+    /// to the ready queue, so it's returned by the service stream.
+    pub fn process(&mut self, transaction: SignedTransaction) -> Result<SignedTransaction> {
+        TxValidator::new(transaction, self.db.clone())
+            .with_all_checks()
+            .validate()
+            .context("tx validation failed")
+            .inspect(|validated_tx| {
+                self.ready_tx.push_back(validated_tx.clone());
+                self.readiness_sender
+                    .send(())
+                    .expect("receiver is always alive");
+            })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TxPoolEvent {
+    PropogateTransaction(SignedTransaction),
+}
+
+impl Stream for TxPoolService {
```

Comment:
@breathx I'm not sure that we really need it in the current design, as after TxPoolService::process returnes successfully `Ok` in the main service `tokio::select` statement, we can just call propagation procedures without any need in this intermediate stream

Reply:
at this moment it seems that we don't need any event here at all, yeah